proposed/swift-interop.md

```diff
@@ -0,0 +1,183 @@
+
+For reference, explicitly declaring a function with the Swift or SwiftAsync calling conventions in Clang requires the "context" argument, the value that goes in the "self" register, as the last parameter or the penultimate parameter followed by the error parameter.
+
+The self register, like the error register and the async context register discussed later, is always a pointer-sized value. As a result, if we introduce any special intrinsic types for the calling convention, we don't need to make the type generic as we can always use a `void*` to represent the value at the lowest level.
+
+We have rejected the following designs for the reasons mentioned below:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+
+Rejected Option 1 seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use a special type like `SwiftError` on a `ref` or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError*` to indicate the error parameter;
+
+Both options characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. The first option represents the "error return" as a managed `ref` or `out` parameter. Like with `SwiftSelf`, we would throw an `InvalidProgramException` for a signature with multiple `SwiftError` parameters.
+
+Option 2 provides an alternative to Option 2 as `ref` and `out` parameters are not supported in `UnmanagedCallersOnly` methods. Option 1 likely provides more possible JIT optimizations as managed pointers are more easily reasoned about in some of our JIT/AOT intermediate representations, but supporting it in all of our goal scenarios (P/Invoke, UnmanagedCallersOnly, function pointers) would require some minor language changes to allow the managed pointers in the `UnmanagedCallersOnly` signature for these special types. Option 2 (`SwiftError*`) does not require any language changes.
```

Comment:
For the `UnmanagedCallersOnly` case, the JIT would need to:
- Introduce a dummy local
- Pretend that the address of this local is method argument to allow user code to set it
- Read the value of the local at the end of the method and store it into the error register

I think it is about at least as complicated to implement as the rejected option 2, for RyuJIT at least. The rejected option 2 may be actually simpler - the JIT just needs to take the SwiftReturn value at the end of the method and store it into the error register and the return value register or return buffer.

Reply:
Sorry for the late response, I just only noticed the design document today. My response here is mainly about RyuJIT, and I realize that UX is a large factor too.

For RyuJIT, there are definitely many separate complexities around each of these. For UnmanagedCallersOnly, with SwiftReturn:
1. For return buffers, the JIT very early transforms these to regular stores. We would still need a separate representation internally to represent the return of the error.
2. For structs returned in registers the JIT can keep struct values in the more natural representation, but this is tied into multi-reg handling and promotion. Some platforms (e.g. win-x64) does not have multi-reg call returns today, so we would need to do the work to support this for the Swift case if we wanted to keep things enregistered. 

I think there are similar complexities in the other direction (when calling a function returning SwiftReturn).

For SwiftError* in the signature you run into another set of problems -- for one, it is not so natural that the parameter/argument differs in a level of indirection to the actual thing we are trying to represent. It also promotes a pattern to the user where they end up using the address of this local in ways that will cause inefficiencies in the codegen.

In the UnmanagedCallersOnly case the suggested implementation by @jkotas sounds reasonable to me. It requires us to special case this dummy local in a number of places because its uses do not appear explicitly in the IR, or otherwise to handle it conservatively (no enregistration). One unfortunate thing here is that it conflates error returns with normal returns -- it will be hard for the JIT to separate these.

Calling something with the SwiftError* is probably the most difficult of all of these cases to represent. If we want to handle this pattern efficiently (without stack spilling) then I do not think this can be represented like this very long within the JIT -- likely we are going to transform something like
```csharp
void* error;
var value = CallToSwiftFunction(x, y, &error);
```
into the RyuJIT IR equivalent of
```csharp
void* error;
var value = CallToSwiftFunction(x, y);
error = SwiftErrorRegister();
```
immediately during importation. Some work will be needed to make sure the CallToSwiftFunction/SwiftErrorRegister pairs do not get separated during JIT optimizations, but I do not think this will be hard. LSRA work will be needed to make sure that the error register is considered busy until it has been read after CallToSwiftFunction.

I might not understand what exactly option 4 entails here, but if it just means having the JIT generate a callout to a VM function when the error register is non-zero after a Swift call, then I think it would be by far the simplest to implement in RyuJIT.

I think option 3 is implementable in a way that does not leave any performance on the table, by considered the operations as intrinsics. I expect that this effectively amounts to directly exposing the mechanisms that RyuJIT will end up with under the hood to the user. I would probably switch "Set" to a direct "return" intrinsic instead, so that we would end up with intrinsics like:
```csharp
[Intrinsic]
void* SwiftErrorReturn() => throw null;
[Intrinsic]
T ReturnSwiftError<T>(void* error) => throw null;
[Intrinsic]
void ReturnSwiftError<T>(void* error) => throw null;
```

SwiftErrorReturn() would return the error returned by the last Swift call. in RyuJIT, it would be transformed directly into a use of the "swift error local" that is defined (by the transformation above) after every call to a Swift function.
The user visible behavior of ReturnSwiftError is an immediate return from the calling function with that error in the error register. A benefit of exposing it like this is that it models that no "normal" return value is necessary (though I am not sure whether or not that is actually how it works in Swift -- can you still look at the value when there was an error return?). Also, it means that we know which returns are error paths, and ensure that these paths to not introduce phantom uses of unrelated values. But the UX of this may not be ideal.

Reply:
Honestly, the shape you mention for the generated code when calling a method is fine with me. Clang uses the same representation.

In any case, we want to use the same representation for calling Swift via P/Invoke and function pointers and UnmanagedCallersOnly.

Our primary focus is on Apple platforms, so the concerns about win-x64 suggests that we should push Swift on Windows support out from v1 if it will have extra costs.

Reply:
> In any case, we want to use the same representation for calling Swift via P/Invoke and function pointers and UnmanagedCallersOnly.

I would say that my main concern for the `SwiftError*` design is around returns in `UnmanagedCallersOnly` functions. In addition to being harder to implement than an intrinsic it promotes writing code that makes it hard for the JIT to separate normal and error returns. For example, returning an error from a function with a large struct return is likely going to incur cost that the same function implemented in Swift wouldn't incur. I am not sure whether it matters or not.

> Our primary focus is on Apple platforms, so the concerns about win-x64 suggests that we should push Swift on Windows support out from v1 if it will have extra costs.

I think with the `SwiftError*` design there shouldn't be much difference in handling between the platforms.

Reply:
> can you still look at the value when there was an error return?

Yes, caller gets a return value and should check for an error immediately after the call. 

> For example, returning an error from a function with a large struct return is likely going to incur cost that the same function implemented in Swift wouldn't incur.

I believe it depends on how heavily `UnmanagedCallersOnly` is utilized by projection tools.

Reply:
> Yes, caller gets a return value and should check for an error immediately after the call.

Is it possible in a Swift function to return both an error and a valid value? Is it possible for the caller to act on both pieces of information, even if there was an error?
For example, when a return buffer is used, I am curious whether the Swift compiler is clearing it out or not on return of an error.

Reply:
Good point. I _think_ the Swift compiler is not clearing out the return buffer if an error is thrown. The caller could still access the return value even though the error register is set. Consider the following example from https://github.com/kotlarmilos/runtime/blob/7bcf61f381fb97edfdf493a4359af89cba8aa4f0/src/tests/Interop/Swift/ErrorHandling.swift#L7-L10.

```swift
public func someFuncThatMightThrow (dummy: UnsafeRawPointer) throws -> Int {
    throw MyError.runtimeError ("Catch me if you can!");
    return 42;
}
```

If the function throws and returns a value, C# caller can still access the return value.

Reply:
CLang/LLVM models swift error the same way as the currently proposed model (pretend that the value is passed via addressable memory): https://clang.llvm.org/docs/AttributeReference.html#swift-error-result , with additional twist that it must be the last parameter (shoud we do the same?). It suggests that the proposed model is a decent choice.

We are going to pay interop transition cost on each of the swift calls. Also, larger structs are typically going to be passed using opaque convention that is expensive compared to passing things in registers. It is ok if the current JIT is not able to 100% optimize some of cases of error results.

Reply:
> Good point. How can we verify that? Probably we can try to set the return register before the call and inspect if it has changed after the call.

I verified it in https://godbolt.org/z/58srMbfK6. `throws_does_not_clear` does not write to the return buffer pointed to by `rax`.

> CLang/LLVM models swift error the same way as the currently proposed model (pretend that the value is passed via addressable memory): https://clang.llvm.org/docs/AttributeReference.html#swift-error-result , with additional twist that it must be the last parameter (shoud we do the same?). It suggests that the proposed model is a decent choice.

As far as I can read in https://github.com/apple/swift/blob/main/docs/ABI/CallingConvention.rst, the Swift compiler lowers Swift signatures into simpler signatures before they are presented to LLVM. It means the Swift compiler can just avoid creating any IR that writes to the return buffer pointer when it sees `throw` expression. I do not think this is an optimization done at the LLVM level.

Reply:
With that said, the design is fine and looks implementable to me, just wanted to point out this difference between `throw` and using a pseudo local.

I am curious whether we are going to do the same kind of lowering as the Swift compiler appears to be doing for signatures. In particular the frozen struct/tuple lowering looks very complex and not described with a lot of detail in https://github.com/apple/swift/blob/main/docs/ABI/CallingConvention.rst. Will the JIT be expected to handle the lowering or can we rely on seeing simplified signatures for swift calls/`UnmanagedCallersOnly`?

Reply:
> CLang/LLVM models swift error the same way as the currently proposed model (pretend that the value is passed via addressable memory): https://clang.llvm.org/docs/AttributeReference.html#swift-error-result , with additional twist that it must be the last parameter (shoud we do the same?)

When calling from C# into Swift, if `error` or `self` are the last parameters, there is no need to handle additional  arguments at the runtime level. These arguments will be automatically ignored by the Swift callee function.